File: go/0110E-平衡二叉树.go
/*
 * @Author: SukiEva
 * @Date: 2021-11-04 09:38:43
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

package main

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func depth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	leftDepth := depth(root.Left)
	rightDepth := depth(root.Right)
	if leftDepth == -1 || rightDepth == -1 || math.Abs(float64(leftDepth)-float64(rightDepth)) > 1 {
		return -1 // 剪枝，只要一个子树不平衡，整个树都不平衡
	}
	if leftDepth > rightDepth {
		return 1 + leftDepth
	} else {
		return 1 + rightDepth
	}
}

func isBalanced(root *TreeNode) bool {
	return depth(root) > -1
}

File: cpp/0110E-平衡二叉树.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-21 10:23:14
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    int maxDepth(TreeNode* root) {
        if (root == nullptr) return 0;
        int leftDepth = maxDepth(root->left);
        int rightDepth = maxDepth(root->right);
        if (leftDepth > rightDepth)
            return leftDepth + 1;
        else
            return rightDepth + 1;
    }
    bool isBalanced(TreeNode* root) {
        if (root == nullptr) return true;
        if (abs(maxDepth(root->left) - maxDepth(root->right)) > 1) return false;
        return isBalanced(root->left) && isBalanced(root->right);
    }
};